src/lib.rs

```diff
@@ -2223,6 +2220,56 @@ pub trait Itertools: Iterator {
         self.collect()
     }
 
+    /// `.product_ok()` is a more convenient way of writing `.product::<Result<_, _>>()`
+    ///
+    /// **Panics** when a primitive integer type is returned and the computation
+    /// overflows, and debug assertions are enabled.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use itertools::Itertools;
+    /// use std::str::FromStr;
+    ///
+    /// fn main() -> Result<(), std::num::ParseIntError> {
+    ///     let product: u64 = ["1", "2", "3"].iter().map(|x| u64::from_str(x)).product_ok()?;
+    ///     assert_eq!(product, 6);
+    ///     Ok(())
+    /// }
+    /// ```
+    fn product_ok<T, U, E>(self) -> Result<U, E>
```

Comment:
Hm, this is an inelegantly large number of generic arguments.

This is a fairly radical alternative, but: What if we just mirrored the stdlib `Try` trait? We can pub-in-priv our `Try`-fork so that folks don't end up explicitly depending on it, and then simply (and potentially non-breaking-ly!) switch over to the standard library `Try` once it's stabilized.

If we do this, our `*_ok` functions will support more than just `Result`s, and they'll be able to take just one generic argument, rather than three (I think).

Thoughts, @Philippe-Cholet and @phimuemue?

Reply:
Something like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7b632d50ccce769af18ec8985ff45359

Reply:
That sure is radical but interesting, maybe it should have its own issue.
About the names, `*_ok` now and `try_*` later? (or `maybe_*`?!)

After a quick look, `ControlFlow` requires 1.55+ while we currently have 1.43. Unless we copy it too?
I previously read something about `#[track_caller]` but don't remember much about its use.

In libcore, `fold` is usually specialized using `try_fold` specializations alongside with `NeverShortCircuit` which surely won't be public forever so I thought that we would copy it at some point.